### PR TITLE
feat: data synced and mapping synced

### PR DESCRIPTION
**Pull request by krackerman**
loopmediagroup/service-search#1910

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -381,15 +381,17 @@ Interacting with the rest api of Elasticsearch
...
 - `data.exists(index: String, id: String)` - check if document exists in alias
+- `data.query(name: String, filter: Object, options: Object)` - query for data in Elasticsearch against alias. Returns raw result body from elasticsearch.
 - `data.refresh(name: String)` - refresh Elasticsearch index, useful e.g. when testing (all versions)
-- `data.update(name: String, options: Object)` - update or delete documents in Elasticsearch (all index versions)
+- `data.signature(index: String, id: String)` - get signature as `${_seq_no')}_${_primary_term}` in alias for document or null if document does not exist
 - `data.stats()` - returns all the statistics for the nodes in a cluster like: indices, cpu usage and other meta
+- `data.synced(index: String)` - returns true if all remote index versions have the same document count 
+- `data.update(name: String, options: Object)` - update or delete documents in Elasticsearch (all index versions)
+- `data.version(index: String, id: String)` - get version number in alias for document or null if document does not exist
```
Comment (krackerman) [inline README.md:394]: These were just not consistent.

File: src/util/rest/rest.js [hunk 2]
```diff
@@ -7,18 +7,20 @@ const mappingGet = require('./mapping/get');
...
-const dataSignature = require('./data/signature');
 const dataExists = require('./data/exists');
 const dataQuery = require('./data/query');
 const dataRefresh = require('./data/refresh');
-const dataUpdate = require('./data/update');
+const dataSignature = require('./data/signature');
 const dataStats = require('./data/stats');
+const dataSynced = require('./data/synced');
+const dataUpdate = require('./data/update');
+const dataVersion = require('./data/version');
```
Comment (krackerman) [inline src/util/rest/rest.js:23]: These were also not consistent.

File: src/util/rest/rest.js [hunk 3]
```diff
@@ -77,17 +79,19 @@ module.exports = (getRels, getMapping, versions, options) => {
...
-      signature: (idx, id) => dataSignature(call, idx, getMapping(idx), id),
       exists: (idx, id) => dataExists(call, idx, id),
       query: (idx, filter) => dataQuery(call, idx, getRels(idx), getMapping(idx), filter),
       refresh: (idx) => dataRefresh(call, idx),
+      signature: (idx, id) => dataSignature(call, idx, getMapping(idx), id),
+      stats: () => dataStats(call),
+      synced: (idx) => dataSynced(call, idx),
       update: (idx, opts) => dataUpdate(call, idx, versions.get(idx), opts),
-      stats: () => dataStats(call)
+      version: (idx, id) => dataVersion(call, idx, getMapping(idx), id)
```
Comment (krackerman) [inline src/util/rest/rest.js:94]: and again

File: src/util/rest/data/synced.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+const count = require('./count');
+const { getIndexVersions } = require('../index/versions');
+
+module.exports = async (call, idx) => {
+  const remoteVersions = await getIndexVersions(call, idx);
```
Comment (simlu) [inline src/util/rest/data/synced.js:5]: we actually want the local versions here

File: src/util/rest/mapping/synced.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const isEqual = require('lodash.isequal');
+const { getIndexVersions } = require('../index/versions');
+
+module.exports = async (call, versions, idx) => {
+  const localVersions = Object.keys(versions.get(idx));
+  const remoteVersions = await getIndexVersions(call, idx);
+  return isEqual(localVersions, remoteVersions);
```
Comment (simlu) [inline src/util/rest/mapping/synced.js:7]: we only want to check that every local version exists remote (since that's what the sync command does)

Comment (simlu) [reply src/util/rest/mapping/synced.js:7]: We might want to have a "pruned" function as well to match what the prune command does?

File: README.md [hunk 6]
```diff
@@ -381,15 +381,17 @@ Interacting with the rest api of Elasticsearch
 - `mapping.list()` - Lists all mappings currently in Elasticsearch
 - `mapping.prune(index: String)` - Removes index versions from Elasticsearch that are not tracked (unknown)
 - `mapping.sync(index: String)` - Creates tracked (known) indices in Elasticsearch when missing
+- `mapping.synced(index: String)` - returns true if local and remote versions are the same
```
Comment (simlu) [inline README.md:384]:
```suggestion
- `mapping.synced(index: String)` - returns true if every local versions exists remotely
```

File: README.md [hunk 7]
```diff
@@ -381,15 +381,17 @@ Interacting with the rest api of Elasticsearch
...
-- `data.version(index: String, id: String)` - get version number in alias for document or null if document does not exist
-- `data.signature(index: String, id: String)` - get signature as `${_seq_no')}_${_primary_term}` in alias for document or null if document does not exist
 - `data.exists(index: String, id: String)` - check if document exists in alias
+- `data.query(name: String, filter: Object, options: Object)` - query for data in Elasticsearch against alias. Returns raw result body from elasticsearch.
 - `data.refresh(name: String)` - refresh Elasticsearch index, useful e.g. when testing (all versions)
-- `data.update(name: String, options: Object)` - update or delete documents in Elasticsearch (all index versions)
+- `data.signature(index: String, id: String)` - get signature as `${_seq_no')}_${_primary_term}` in alias for document or null if document does not exist
 - `data.stats()` - returns all the statistics for the nodes in a cluster like: indices, cpu usage and other meta
+- `data.synced(index: String)` - returns true if all remote index versions have the same document count
```
Comment (simlu) [inline README.md:392]:
```suggestion
- `data.synced(index: String)` - returns true if all local index version exists on remote and have the same document count 
```

**Review by simlu — APPROVED** [review]

**Pull request merged**